src/lightning/data/streaming/dataset.py

```diff
@@ -11,36 +11,135 @@
+
+        self.worker_chunks: List[int] = []
+        self.worker_intervals: List[List[int]] = []
+        self.current_indexes: List[int] = []
+        self.chunk_index = 0
+        self.index = 0
+        self.has_triggered_download = False
+        self.min_items_per_replica: Optional[int] = None
+        self.seed = seed
+        self.num_iter = 0
+        self.random_state = None
 
     def __len__(self) -> int:
-        return len(self.cache)
+        return self.L
+
+    def __iter__(self) -> "StreamingDataset":
+        self.random_state = np.random.RandomState(seed=self.seed + self.num_iter)  # type: ignore
+        chunk_intervals = self.cache.get_chunk_interval()
+        indexes = range(len(chunk_intervals))
+        shuffled_indexes = self.random_state.permutation(indexes) if self.shuffle else list(indexes)
+        shuffled_chunk_intervals = np.asarray(chunk_intervals)[shuffled_indexes]
+
+        chunks_per_replica: List[List[int]] = [[] for _ in range(self.distributed_env.world_size)]
+        intervals_per_replica: List[List[List[int]]] = [[] for _ in range(self.distributed_env.world_size)]
+        for index, (chunk_index, chunk_interval) in enumerate(zip(shuffled_indexes, shuffled_chunk_intervals)):
+            replica_index = index % self.distributed_env.world_size
+            chunks_per_replica[replica_index].append(chunk_index)
+            intervals_per_replica[replica_index].append(chunk_interval)
+
+        current_chunks = chunks_per_replica[self.distributed_env.global_rank % self.distributed_env.world_size]
+        current_intervals = intervals_per_replica[self.distributed_env.global_rank % self.distributed_env.world_size]
+
+        if self.worker_env is None:
+            self.worker_env = _WorkerEnv.detect()
+
+        self.worker_chunks = []
+        self.worker_intervals = []
+
+        for i, (chunk_index, chunk_interval) in enumerate(zip(current_chunks, current_intervals)):
+            if i % self.worker_env.world_size != self.worker_env.rank:
+                continue
+            self.worker_chunks.append(chunk_index)
+            self.worker_intervals.append(chunk_interval)
+
+        self.current_indexes = []
+        self.chunk_index = 0
+        self.num_iter += 1
+
+        return self
+
+    def __getitem__(self, index: Union[ChunkedIndex, int]) -> Any:
+        if isinstance(index, int):
+            index = ChunkedIndex(index, self.cache._get_chunk_index_from_index(index))
+        return self.cache[index]
+
+    def __next__(self) -> Any:
+        # Lazily re-populate the interval to reduce memory usage.
+        if len(self.current_indexes) == 0:
+            if self.chunk_index == len(self.worker_intervals):
+                raise StopIteration
+
+            interval = self.worker_intervals[self.chunk_index]
+            current_indexes = np.arange(0, interval[1] - interval[0])
+            if self.shuffle:
+                current_indexes = self.random_state.permutation(current_indexes)
```

Comment:
do we need to determine how we permute the indices ahead of time to make this fault tolerant? Or we just keep the `random_state` around to eventually replay back?

Reply:
Yes, this is left for another PR.